### ISSUE TITLE
Feature: Ability to know when Websocket connected successfully (i.e. optional, user-defined event handlers for lifecycle visibility)

### ISSUE THREAD

**Issue by defidefi5** [nktkas/hyperliquid#6]
**Motivation and Use Case**

With a regular web socket class, I'd define an 'open' event handler and invoke resolve within it (see example).

This allows me to use await with `connectToFeed()` _in a meaningful way_ because the `connectToFeed()` promise only resolves when the web socket is established successfully.

```ts
  // within my class
  connectToFeed(): Promise<void> {
    return new Promise((resolve, reject) => {
      
      const connectionTimeout = setTimeout(() => {
        reject(new Error("WebSocket connection timeout (5s)"));
      }, 5000);

      this.ws = new WebSocket("...");
      this.ws.addEventListener("message", (message) => {
        console.info(message);
      });
      this.ws.addEventListener("open", () => {
        console.info("WS connected");
        clearTimeout(connectionTimeout);
        resolve();  // <<<<<<<<<<<<<<<<< here
      });
      this.ws.addEventListener("close", () => {
        console.info("WS closed");
      });
      this.ws.addEventListener("error", (error) => {
        console.error("WS error", error);
        reject(error);
      });
    });
  }
  ```

**Current behavior**

However, this library does not allow the end user to know when the websocket connection was established successfully.

```ts
  // within my class
  connectToWSPublic(): Promise<void> {
    return new Promise((resolve, reject) => {
      const publicTransport = new hl.WebSocketTransport({
        url: "wss://api.hyperliquid.xyz/ws",
        timeout: 10_000,
      });
  
      this.wsPublic = new hl.PublicClient(publicTransport);

      // Not a valid way to use resolve b/c it does not wait for a successful websocket
      // connection before resolving.
      resolve();
    });
  }
  ```

## Possible solution

Allow optional, user-defined callback handlers for lifecycle events (the transport object can run its existing logic for these handlers, but also check if the defined callback function exists--using `typeof onOpen === 'function'`--and invoke it if defined.)

Which leads me to then think that user-defined `onOpen`, `onClose`, and `onError` events are probably the solution b/c this solves a related issue, which is the library does provide visibility into the lifecycle of the web socket connection, which would achieved by allowing a user to define custom callbacks and add their desired console log, or whatever they may need, there.

(I did not mention `onMessage`, b/c for the PublicClient and WalletClient, their expected behavior is of the library to handle 'message' events; but your upcoming web socket event subscription class might be solved just by user to define their own `send()` method and `onMessage()` handler for the reconnecting class.)

**Comment by nktkas** [comment]
I agree, I missed that point. Background initialization of WebSocket may not suit everyone. Also, because of the "protected" variable modifier, you can't interact with the socket from under the transport without ignoring the TypeScript warning.

I propose to add a new method in WebSocketTransport "ready" which waits for connection opening, and also change the modifier of socket variable from "protected" to "readonly". I consider proxying "onOpen", "onClose" and "onError" from socket to WebSocketTransport unnecessary.

**Comment by defidefi5** [comment]
yeah, `onOpen` or `ready` is most important. 

To illustrate the use case for the others, `onClose` or `onError` would allow an app to handle the lifecycle in different ways, like disabling trading until reconnected or sending a message to a a database for visibility into the app state. For now, I can live without these, but can see how they might be useful.

**Comment by nktkas** [comment]
Done at https://github.com/nktkas/hyperliquid/commit/6cdfaa4a66d148357f37a4961ca83c7282fbf14a.

How to wait for a connection WebSocketTransport:
```ts
import { PublicClient, WebSocketTransport } from "@nktkas/hyperliquid";

const transport = new WebSocketTransport();
const client = new PublicClient({ transport });
await client.transport.ready(); // or `await transport.ready();`
```

How to add custom listeners to WebSocketTransport:
```ts
import { PublicClient, WebSocketTransport } from "@nktkas/hyperliquid";

const transport = new WebSocketTransport();
const client = new PublicClient({ transport });

client.transport.socket.addEventListener("open", () => {
    console.log("Connected to the server");
});
client.transport.socket.addEventListener("close", () => {
    console.log("Disconnected from the server");
});

await client.transport.ready();
await client.transport.close();
```
Note that WebSocketTransport has a default reconnect function and added listeners without the `{ once: true }` modifier will be re-run.

**Comment by defidefi5** [comment]
Cool! Thanks
